### PR TITLE
RATIS-1778. Fix NPE in readIndexHeartbeats

### PR DESCRIPTION
**Pull request by codings-dan**
see https://issues.apache.org/jira/browse/RATIS-1778

### CODE REVIEW

File: ratis-server/src/main/java/org/apache/ratis/server/impl/ReadIndexHeartbeats.java [hunk 1]
```diff
@@ -101,7 +101,7 @@ boolean receive(LogAppender logAppender, AppendEntriesReplyProto proto,
       final HeartbeatAck reply = replies.computeIfAbsent(
           logAppender.getFollowerId(), key -> new HeartbeatAck(logAppender));
       if (reply.receive(proto)) {
-        if (hasMajority.test(id -> replies.get(id).isAcknowledged())) {
+        if (hasMajority.test(id -> replies.get(id) != null && replies.get(id).isAcknowledged())) {
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/ReadIndexHeartbeats.java:104]: Let's use `Optional` to add an `isAcknowledged` method:
```java
    boolean isAcknowledged(RaftPeerId id) {
      return Optional.ofNullable(replies.get(id)).map(HeartbeatAck::isAcknowledged).isPresent();
    }
```

Comment (codings-dan) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/ReadIndexHeartbeats.java]: done

**Comment by codings-dan** [comment]
@szetszwo Thanks a lot for reviewing the pull request!

File: ratis-server/src/main/java/org/apache/ratis/server/impl/ReadIndexHeartbeats.java [hunk 2]
```diff
@@ -110,6 +111,10 @@ boolean receive(LogAppender logAppender, AppendEntriesReplyProto proto,
       return isCompletedNormally();
     }
 
+    boolean isAcknowledged(RaftPeerId id) {
+      return Optional.ofNullable(replies.get(id)).map(HeartbeatAck::isAcknowledged).isPresent();
```
Comment (kaijchen) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/ReadIndexHeartbeats.java:115]: What if `replies.get(id)` is not null, but `HeartbeatAck::isAcknowledged` returns false?

```java
      return Optional.ofNullable(replies.get(id)).map(HeartbeatAck::isAcknowledged).getOrElse(false);
```
```java
      return Optional.ofNullable(replies.get(id)).filter(HeartbeatAck::isAcknowledged).isPresent();
```

Comment (codings-dan) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/ReadIndexHeartbeats.java:115]: Thanks for helping review the pull request! The current changes have introduced new problems and your two changes are correct. Let me revert this and then submit a new pull request!

**Comment by codings-dan** [comment]
Retrying this on https://github.com/apache/ratis/pull/819

**Pull request merged**